MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -352,6 +356,342 @@ void GeoMapper::advancedMapOnMesh(
+    GeoLib::LineSegment segment)
+{
+    // modify z coordinates such that all surface elements around the line
+    // segment are found
+    segment.getBeginPoint()[2] = mesh_element_grid.getMinPoint()[2];
+    segment.getEndPoint()[2] = mesh_element_grid.getMaxPoint()[2];
+    std::array<MathLib::Point3d, 2> const pnts{
+        {segment.getBeginPoint(), segment.getEndPoint()}};
+    GeoLib::AABB aabb(pnts.cbegin(), pnts.cend());
+
+    auto candidate_elements = mesh_element_grid.getElementsInVolume(
+        aabb.getMinPoint(), aabb.getMaxPoint());
+
+    // make candidate elements unique
+    BaseLib::makeVectorUnique(candidate_elements);
+
+    return candidate_elements;
+}
+
+static bool snapPointToElementNode(MathLib::Point3d& p,
+                                   MeshLib::Element const& elem, double eps)
+{
+    for (std::size_t k(0); k<elem.getNumberOfNodes(); ++k) {
+        double const sqr_dist_2d(MathLib::sqrDist2d(p, *(elem.getNode(k))));
+        if (sqr_dist_2d < eps) {
+#ifndef NDEBUG
+            std::stringstream out;
+            out.precision(std::numeric_limits<double>::digits10);
+            out << "Segment point snapped from " << p;
+#endif
+            p[0] = (*(elem.getNode(k)))[0];
+            p[1] = (*(elem.getNode(k)))[1];
+            p[2] = (*(elem.getNode(k)))[2];
+#ifndef NDEBUG
+            out << "to " << p;
+            DBUG("%s", out.str().c_str());
+#endif
+            return true;
+        }
+    }
+    return false;
+}
+
+static void insertSubSegments(
+    GeoLib::Polyline& ply, GeoLib::PointVec& points,
+    GeoLib::Polyline::SegmentIterator& segment_it,
+    std::vector<GeoLib::LineSegment> const& sub_segments)
+{
+    std::size_t const j(segment_it.getSegmentNumber());
+    std::size_t new_pnts_cnt(0);
+    for (auto const& segment : sub_segments)
+    {
+        auto const begin_id(points.push_back(
+            new GeoLib::Point(segment.getBeginPoint(), points.size())));
+        if (ply.insertPoint(j + new_pnts_cnt + 1, begin_id))
+            new_pnts_cnt++;
+        auto const end_id(points.push_back(
+            new GeoLib::Point(segment.getEndPoint(), points.size())));
+        if (ply.insertPoint(j + new_pnts_cnt + 1, end_id))
+            new_pnts_cnt++;
+    }
+    segment_it += new_pnts_cnt;
+}
+
+static void mapPolylineOnSurfaceMesh(
+    GeoLib::Polyline& ply,
+    GeoLib::PointVec& orig_points,
+    MeshLib::MeshElementGrid const& mesh_element_grid)
+{
+    // for each segment ...
+    for (auto segment_it(ply.begin()); segment_it != ply.end(); ++segment_it)
+    {
+        auto candidate_elements(getCandidateElementsForLineSegmentIntersection(
+            mesh_element_grid, *segment_it));
+
+        auto mapPoint = [&candidate_elements](MathLib::Point3d& p) {
+            auto const* elem(
+                findElementContainingPoint2d(candidate_elements, p));
+            if (elem)
+            {
+                if (!snapPointToElementNode(p, *elem, 1e-3))
```

Comment:
1. The *squared* distance have to smaller than `1e-3`, i.e., the distance has to be smaller than `1e-6`. I hope this is acceptable in your applications?
2. Since in adaptive meshes the "mesh density" can vary a lot, there can not be a single value for this tolerance. So far I haven't any idea to solve this issue in general.
3. In this particular case a possible solution would be to use the mesh element size.